suite/40-mrouter.sh

```diff
@@ -77,104 +104,138 @@ mrouter_report_one()
 mrouter_inject_and_report()
 {
     local all_grps="$1"
-    local b2_grps="$2"
-    local b3_grps="$3"
-    local b3_mrouter=$4
+    local out_grps="$2"
+    local rport_grps="$3"
+    local rport_mrouter=$4
 
-    capture -f "icmp or icmp6 or ether proto 0xbbbb" $h2 $h3
+    capture -f "icmp or icmp6 or ether proto 0xbbbb" $oh $rh
 
-    step "  Inject MAC/IPv4/IPv6 multicast to all groups ($all_grps) on $h1"
+    step "  Inject MAC/IP multicast to all groups ($all_grps) on $ih"
     for g in $all_grps; do
-	eth -g $g | { cat; echo U from $h1; } | inject $h1
-	mcast_gen -c 1 -g $(ip4_grp $g) $ip4_src
-	mcast_gen -c 1 -g $(ip6_grp $g)%$h1 $ip6_src
+	eth -g $g | { cat; echo U from $ih; } | inject $ih
+	mcast_gen -c 1 -g $(mrouter_ip4_grp $g) $mrouter_ip4_src
+
+	# TODO: Nemesis can't yet generate MLD queries, so we only
+	# test IPv6 for the port version of the test, i.e. when the
+	# bridge is generating queries internally.
+	if [ $mode = port ]; then
+	    mcast_gen -c 1 -g $(mrouter_ip6_grp $g)%$ih $mrouter_ip6_src
+	fi
     done
 
-    mrouter_report_one $h2 no "$b2_grps" "$all_grps" || return 1
-    mrouter_report_one $h3 $b3_mrouter "$b3_grps" "$all_grps" || return 1
+    mrouter_report_one $oh no "$out_grps" "$all_grps" || return 1
+    mrouter_report_one $rh $rport_mrouter "$rport_grps" "$all_grps" || return 1
 }
 
-mrouter_port()
+mrouter_test()
 {
     require3loops
 
-    local in=$b1
-    local out=$b2
-    local rport=$b3
+    mode=$1
 
-    local bropts="mcast_snooping 1"
+    ib=$b1
+    ih=$h1
+    ob=$b2
+    oh=$h2
+    qb=$([ $mode = host ] && echo $b3 || echo $br0)
+    qh=$([ $mode = host ] && echo $h3)
+    rb=$([ $mode = port ] && echo $b3 || echo $br0)
+    rh=$([ $mode = port ] && echo $h3 || echo $br0)
```

Comment:
Perhaps add a comment explaining the meanings of `i`, `o`, `q`, `r`.

Reply:
Good point! I'll add a legend with descriptions.